feat: add SVGO optimization support for SVG assets

azat-io opened the pull request:
## Changes

- Added `svg.optimize` and `svg.svgoConfig` configuration options
- Integrated SVGO optimization in the SVG processing pipeline
- Updated type definitions and config schema
- Added error handling with fallback to original SVG content

## Testing

I'll add tests in the next commit.

## Benefits

- Reduces SVG file sizes automatically during build
- Configurable optimization with full SVGO plugin support
- Backward compatible - optimization is opt-in
- Graceful error handling with fallback

## Docs

I'll create another PR to docs.

delucis commented on packages/astro/src/core/config/schemas/base.ts:
```diff
@@ -280,6 +284,12 @@ export const AstroConfigSchema = z.object({
+	svg: z
+		.object({
+			optimize: z.boolean().default(true),
+			svgoConfig: z.record(z.any()).optional(),
```
I think we should offer at least proper typing here even if we don’t do full schema validation. For example, if a type is available for the SVGO options, we could use something like:
```js
			svgoConfig: z.custom<SVGOOptions>((value) => value && typeof value === 'object').optional(),
```
That way users should get a nice editor experience.

delucis commented on packages/astro/src/types/public/config.ts:
```diff
@@ -1379,6 +1380,54 @@ export interface ViteUserConfig extends OriginalViteUserConfig {
+		 * @description
+		 * Whether to enable SVG optimization using SVGO during build time.
+		 *
+		 * When enabled, all imported SVG files will be optimized for smaller file sizes
```
> all imported SVG files

Will this also apply to usage with Astro’s assets pipeline? For example usage like:

```astro
---
import { Image } from 'astro:assets';
import mySVG from './example.svg';
---

<Image src={mySVG} alt="Example SVG" />
```

delucis reviewed (commented):
Thanks for the PR @azat-io! I think this could be a nice feature.

Left a couple of questions based on a quick look.

Also pinging @natemoo-re and @stramel as our resident SVG experts who might have good insights here!

azat-io commented:
Made fixes based on comments:

- added changest
- improved typing
- made the feature experimental

ascorbic commented on packages/astro/src/types/public/config.ts:
```diff
@@ -2524,6 +2525,54 @@ export interface ViteUserConfig extends OriginalViteUserConfig {
+		svg?: {
+			/**
+			 * @docs
+			 * @name svg.optimize
```
```suggestion
			 * @name experimental.svg.optimize
```

azat-io replied:
Done ✅

ascorbic commented on packages/astro/src/types/public/config.ts:
```diff
@@ -2524,6 +2525,54 @@ export interface ViteUserConfig extends OriginalViteUserConfig {
+
+			/**
+			 * @docs
+			 * @name svg.svgoConfig
```
```suggestion
			 * @name experimental.svg.svgoConfig
```

azat-io commented:
Should I create a pull request in the documentation?

azat-io commented:
Rebased.

azat-io commented:
Ping.

azat-io commented:
Hi @delucis, @stramel, @natemoo-re,

Checking in - it's been 4 months since the last comment. All feedback has been addressed and the PR is ready for review.

What's the status on this?

Princesseuh commented on packages/astro/src/assets/vite-plugin-assets.ts:
```diff
@@ -244,7 +245,13 @@ export default function assets({ fs, settings, sync, logger }: Options): vite.Pl
+							code: makeSvgComponent(
+								imageMetadata,
+								contents,
+								settings.config.experimental?.svg as AstroConfigType['experimental']['svg'],
```
A bit confused about this cast? Is it really necessary?

azat-io commented on .changeset/cyan-tables-poke.md:
```diff
@@ -0,0 +1,5 @@
+---
+'astro': patch
```
Fixed.

azat-io commented on .changeset/cyan-tables-poke.md:
```diff
@@ -0,0 +1,5 @@
+'astro': patch
+---
+
+Add experimental SVGO optimization support for SVG assets
```
Fixed

azat-io commented:
I've made fixes based on your comments, rebased, and resolved conflicts. I also opened a new pull request for documentation:

https://github.com/withastro/docs/pull/12637

sarah11918 commented on .changeset/cyan-tables-poke.md:
```diff
@@ -0,0 +1,5 @@
+'astro': minor
+---
+
+Adds experimental SVGO optimization support for SVG assets
```
This is indeed the verb tense we use. Good Robot! 😄 

@azat-io I'll point you to our guidance for [writing changesets to introduce experimental features](https://contribute.docs.astro.build/docs-for-code-changes/experimental-feature-docs/#what-to-include-in-the-changeset) so we can hype up this feature a bit!

For minor releases, which typically are new features, we also allow a bit more kind of "promoting" of the feature here, so people know whether or not they want to try it. Experimental features are great because they let us test and find bugs early, but typically, only a small number of users actually enable them, so we lose out on that if we don't make your feature sound worth the risk! 😉 

The docs above show what we would typically write here, so can you try to model something off of that? Here is some examples of a recent experimental feature changeset you can also use for reference:  

- #13084 
- #13402 
- #13971 

Note too that if your feature replaces/supercedes/is incompatible with some project code people may be using to achieve the same result on their end, we do mention e.g. "If you were doing X to achieve this result, you may need to inspect and update your project code to remove Y" (I don't know if there is some SVG optimization people might have written for themselves, or workarounds they would have used that this can replace)

Would you be up for taking a shot at writing a changeset of that style? We'll all certainly help get it over the finish line!

azat-io replied:
✅️️️️

sarah11918 commented on packages/astro/src/types/public/config.ts:
```diff
@@ -2537,6 +2538,54 @@ export interface AstroUserConfig<
 		chromeDevtoolsWorkspace?: boolean;
+
+		/**
+		 * @docs
```
We would typically not include experimental config options in the docs reference page (everything under top-level `experimental` is kept to its own experimental page), so I don't think any of the `@docs` entries are needed here.

azat-io replied:
✅️️️️

sarah11918 reviewed (commented):
Some comments on behalf of docs here, and thank you for the docs PR @azat-io !

1. Quick question (for everyone) on the experimental flag name. `svg` is pretty general, and could mean anything *about* SVGs. (And, could create confusion if there's another experimental feature involving SVGs at the same time.) So just want to make sure this shouldn't be something like `experimental.svgOptimization` for clarity, and to let the user know exactly what they're signing up for!

2. Also, wondering about the `optimize` property which as far as I can tell is basically just setting the flag true? This doesn't seem to match our other patterns that can be just enabled, or passed extra config, like CSP for example. You either set the experimental flag true to enable the basic behaviour, or you pass a config object (but you don't otherwise have a property that retains "turning on the flag" behaviour):

```
export default defineConfig({
  experimental: {
    csp: {
      directives: [
        "default-src 'self'",
        "img-src 'self' https://images.cdn.example.com"
      ]
    }
  }
});
```

It feels weird to document "optimize" when it really just means "yes, use this flag" and is required for any of the other options to be set. (Which otherwise are just one object that could even be `experimental.svg` options directly?) Maybe this is future proofing for when it's not experimental and these will be higher-level options? But it feels clunkier than other patterns we've used as I'm looking at documenting it, so I'm just pointing it out because I'm noticing it!

azat-io commented:
@sarah11918 @florian-lefebvre 

Thanks for another iteration of the review.

What has been done:

1. Improved changeset, added documentation
2. Removed `@docs` from the interface
3. Fixed `AstroConfigType` to `AstroConfig`
4. Fixed merge issues
5. Fixed default behavior to `false`

azat-io commented on packages/astro/src/core/config/schemas/base.ts:
```diff
@@ -106,6 +107,10 @@ export const ASTRO_CONFIG_DEFAULTS = {
 		chromeDevtoolsWorkspace: false,
 		failOnPrerenderConflict: false,
+		svg: {
+			optimize: true,
```
✅️️️️

florian-lefebvre replied:
Created https://github.com/withastro/astro/issues/14691 to track

azat-io commented on packages/astro/src/assets/utils/svg.ts:
```diff
@@ -1,10 +1,24 @@
+import { optimize } from 'svgo';
+
 import { parse, renderSync } from 'ultrahtml';
+import type { AstroConfigType } from '../../core/config/schemas/index.js';
```
✅️️️️

florian-lefebvre commented on packages/astro/src/assets/utils/svg.ts:
```diff
@@ -1,10 +1,24 @@
+			const result = optimize(contents, svgConfig.svgoConfig);
+			processedContents = result.data;
+		} catch (error) {
+			console.warn('SVGO optimization failed:', error);
```
Sure! You need to do 2 things basically:
1. Create a new error data in `src/core/errors/error-data.ts` (path might not be exact but pretty close). You can look at `src/core/errors/README.md` for guidance and at the rest of the TS file for practical examples
2. Then here you can throw an AstroError directly, something like this:

```ts
catch(cause) {
	throw new AstroError(
		AstroErrorData.SvgoOptimizationFailed,
		{ cause }
	)
}
```
Let me know if you need any clarification

florian-lefebvre commented on packages/astro/src/assets/utils/svg.ts:
```diff
@@ -1,10 +1,24 @@
+
 import { parse, renderSync } from 'ultrahtml';
+import type { AstroConfig } from '../../types/public/config.js';
+
```
```suggestion
```

florian-lefebvre commented on packages/astro/src/assets/utils/svg.ts:
```diff
@@ -1,10 +1,24 @@
-	const root = parse(contents);
+function parseSvg(contents: string, svgConfig?: AstroConfig['experimental']['svg']) {
+	let processedContents = contents;
+	if (svgConfig?.optimize) {
```
```suggestion
	if (svgConfig.optimize) {
```

florian-lefebvre commented on packages/astro/src/assets/utils/svg.ts:
```diff
@@ -17,9 +31,13 @@ function parseSvg(contents: string) {
+export function makeSvgComponent(
+	meta: ImageMetadata,
+	contents: Buffer | string,
+	svgConfig?: AstroConfig['experimental']['svg'],
```
```suggestion
	svgConfig: AstroConfig['experimental']['svg'],
```

azat-io replied:
Done ✅️️️️

florian-lefebvre commented on packages/astro/src/assets/vite-plugin-assets.ts:
```diff
@@ -272,7 +272,9 @@ export default function assets({ fs, settings, sync, logger }: Options): vite.Pl
 							// We know that the contents are present, as we only emit this property for SVG files
-							return { code: makeSvgComponent(imageMetadata, contents) };
+							return {
+								code: makeSvgComponent(imageMetadata, contents, settings.config.experimental?.svg),
```
```suggestion
								code: makeSvgComponent(imageMetadata, contents, settings.config.experimental.svg),
```

azat-io replied:
Done ✅️️️️

florian-lefebvre commented on packages/astro/src/core/config/schemas/base.ts:
```diff
@@ -106,6 +107,10 @@ export const ASTRO_CONFIG_DEFAULTS = {
 		failOnPrerenderConflict: false,
+		svg: {
+			optimize: false,
+			svgoConfig: {},
```
Unused
```suggestion
```

florian-lefebvre reviewed (commented):
Little knits

sarah11918 commented on .changeset/cyan-tables-poke.md:
````diff
@@ -0,0 +1,38 @@
+});
+```
+
+The feature includes graceful error handling - if optimization fails for any SVG, Astro will fall back to using the original file content.
````
```suggestion
The feature includes graceful error handling. If optimization fails for any SVG, Astro will fall back to using the original file content.
```

azat-io replied:
Done ✅️️️️

sarah11918 commented on .changeset/cyan-tables-poke.md:
```diff
@@ -0,0 +1,38 @@
+
+The feature includes graceful error handling - if optimization fails for any SVG, Astro will fall back to using the original file content.
+
+For more information on enabling and using this feature in your project, see the [Experimental SVG Optimization docs](https://docs.astro.build/en/reference/experimental-flags/svg/).
```
```suggestion
For more information on enabling and using this feature in your project, see the [experimental SVG optimization docs](https://docs.astro.build/en/reference/experimental-flags/svg-optimization/).
```

azat-io replied:
Done ✅️️️️

azat-io commented on .changeset/cyan-tables-poke.md:
````diff
@@ -0,0 +1,38 @@
+});
+```
+
+You can customize SVGO behavior by providing your own configuration:
````
Done ✅️️️️

azat-io commented on packages/astro/src/assets/utils/svg.ts:
```diff
@@ -1,10 +1,24 @@
 
-function parseSvg(contents: string) {
-	const root = parse(contents);
+function parseSvg(contents: string, svgConfig?: AstroConfig['experimental']['svg']) {
```
Done ✅️️️️

azat-io commented on packages/astro/src/assets/utils/svg.ts:
```diff
@@ -1,10 +1,24 @@
+import { optimize } from 'svgo';
+
```
Done ✅️️️️

sarah11918 commented on packages/astro/src/core/errors/errors-data.ts:
```diff
@@ -2064,6 +2064,15 @@ export const SessionConfigWithoutFlagError = {
+/**
+ * @docs
+ */
+export const CannotOptimizeSvg = {
```
Some notes here:

- This error message should probably tell people what is going to happen and/or what they should do. I believe from the feature it will just fall back to an unoptimized SVG and there's nothing they can actually do about it? So, probably makes sense to additionally say that the original SVG will be used/building with the original SVG instead... something like that.
- I know sometimes we can say *which* file led to the error. If so, it's a nicety here to be able to say which SVG could not be optimized. This lets people try to do something with that file in particular, if they want. And, in that case, you could suggest optimizing it manually through something like SVGOMG, if desired, in the error message.
- These errors are all [indexed on a single docs page](https://docs.astro.build/en/reference/error-reference/) in the order that they're listed here, and they are in categories. So, if there's a better category/placement, then we add new errors in that location in the file. We can't just stick new ones at the end or they end up in the last category. Right now the existing image errors are in the catch-all "Astro Errors" category, so it probably makes sense to put it in that one. (We could and maybe should have a separate images error section, but that's not the responsibility of this PR!)

florian-lefebvre replied:
Updated in [`3b067d5` (#13880)](https://github.com/withastro/astro/pull/13880/commits/3b067d511f3a894bd4ccf3407ebb02ac15406671), PTAL!

florian-lefebvre commented on packages/astro/src/core/errors/errors-data.ts:
```diff
@@ -2066,11 +2066,13 @@ export const SessionConfigWithoutFlagError = {
 	title: 'Cannot optimize SVG',
-	message: 'An error occurred while optimizing a SVG file with SVGO.',
+	message: (path: string) => `An error occurred while optimizing SVG file "${path}" with SVGO.`,
+	hint: 'Review the error cause to fix this error.',
```
I agree it wasn't a better error, or just slightly xd. But seriously:

> which I believe is your project will build but using an unoptimized SVG

No, if the optimization fails the build fails with this error. We can revisit this if you want by only logging a warning as it was originally

> Do you tell me what the cause is

The reason I added this hint is because when we throw the `AstroError`, we pass this error data + the original error from SVGO (which shows below in the terminal or overlay). That's what I meant but I agree it's not very clear, I didn't know how to explain that.

That's why I can't really be very specific here, because I don't know! SVGO knows

florian-lefebvre replied:
Only cases 1 and 3 exist ATM. I don't think we want 2 **and** 3 but there's probably a discussion to be had about whether we want 2 **or** 3. Any opinion on this?

sarah11918 replied:
2 **and** 3 would be problematic, I think 😅 I pointed out the line in the changeset that led to my faulty thinking.

sarah11918 commented on .changeset/cyan-tables-poke.md:
````diff
@@ -0,0 +1,40 @@
+});
+```
+
+The feature includes graceful error handling. If optimization fails for any SVG, Astro will fall back to using the original file content.
````
@florian-lefebvre This is what I mean re: Case 2. What does this mean wrt getting errors optimizing?

(And then, similarly in the Docs PR here https://github.com/withastro/docs/pull/12637/files#r2489752607 )

florian-lefebvre replied:
This was the original behavior but this got changed after my feedback, I believe.

sarah11918 commented on .changeset/cyan-tables-poke.md:
````diff
@@ -0,0 +1,40 @@
+});
+```
+
+The feature includes graceful error handling. If optimization fails for any SVG, Astro will fall back to using the original file content.
````
```suggestion
```
No longer true! I think we can just delete.

sarah11918 reviewed (approved):
Approving for docs!
